### DISCUSSION TITLE
How can I start the server on a specific port

### DISCUSSION THREAD

**Discussion by krystalcode** [haskell/haskell-language-server#4468] [Q&A] [answered]
I am trying to set up the Haskell language server and connect it to my Emacs. I want to run the server within a container and connect it to my Emacs with networking. I don't see any options for defining a port for the server to listen to, is this not supported?

How does it currently work, via STDIO?

**Comment by fendor** [comment]
Hi!

Yes, LSP servers by default read messages from stdin and send messages via stdout.

**Reply by krystalcode** [reply]
Thank you for the response.

This makes it more complicated to make it work with containers. Is starting the server on a port something that could be supported i.e. create an issue for it? Maybe if there's guidance I can try implementing this.

If it would be preferred to not add this to this package, what could be an alternative approach? Maybe a separate utility server that relays stdin/stdout through a port - maybe there's something already existing for that?

**Reply by fendor** [reply]
Depending on your setup, it might make sense to bundle the LSP with your build/development environment, as the language server is essentially a compiler, and requires access to `cabal` and `ghc`, etc... Feels like there would be a lot of duplication.

I think enabling HLS to listen and send messages to a port isn't a lot of work, as we essentially only need a `Handle`. But you'd need to touch multiple files.

> Maybe a separate utility server that relays stdin/stdout through a port - maybe there's something already existing for that?

I think that's the easiest solution if a separate container for HLS is necessary. I can imagine that there is such a separate utility, but I am not aware of one, yet.

**Comment by krystalcode** [comment, marked as answer]
For anybody that wants to do something similar, I figured out how to make it work via STDIO using `socat`:

`socat tcp-listen:8893,reuseaddr exec:"haskell-language-server-wrapper --lsp"`

Here is an example entrypoint for a container: https://github.com/krystalcode/docker-dev/blob/master/language-servers/haskell/entrypoint.sh
